stm32cubemx: fix broken package updates

mrubli opened the pull request:
This fixes two subtle but nasty bugs with STM32CubeMX 6.14:

1. Logging into myST – a prerequisite for downloading packages – does not work. Clicking the corresponding button in the _Connections & Updates_ dialog simply does nothing. The only workaround is to copy config files from a system on which this works.
2. Software package index updates do not work. They appear to download just fine, but the list of available packages remains unchanged.

The details are described in the individual commit messages.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

angaz commented:
Hi @mrubli, someone else made a PR to fix the problem with udev, it's a bit more accurate than including systemd, so can you please drop that commit.
I have a completely white window when the JxBrowser is showing, so I can't test anything because it just doesn't work for me no matter what I do. Your fix for updating the packages seems good though. So I will accept that.

mrubli commented:
Hi @angaz. Thanks for reviewing and for pointing me at that PR! I had originally tried something like that but without success, so I'm glad someone figured out a more elegant solution. :+1: 

I've dropped that commit from the PR and rebased my branch on top of the latest `master` branch.

angaz commented:
Yeah I was able to reproduce the problem and it seem this helps. Thanks!

pbsds commented on pkgs/by-name/st/stm32cubemx/package.nix:
```diff
@@ -52,6 +52,11 @@ let
+      updater_xml="\$HOME/.stm32cubemx/thirdparties/db/updaterThirdParties.xml"
+      if [ -e "\$updater_xml" ] && [ ! -w "\$updater_xml" ]; then
+        echo "Warning: Unwritable \$updater_xml prevents CubeMX software packages from working correctly. Fixing that."
+        chmod u+w "\$updater_xml"
```
```suggestion
        (set -x; chmod u+w "\$updater_xml")
```

mrubli replied:
I've applied your suggestion and rebased the branch on top of `master` again.

mrubli committed:
stm32cubemx: fix broken software package database update

For some reason, STM32CubeMX initially creates one of its database files with
missing user-write permissions. When this happens, refreshing the package list
in the Embedded Software Packages Manager seems to succeed but new packages
don't become available. (As of the time of writing, this is the case with
TouchGFX 4.25.0, which fails to appear, and 4.24.2 showing as the latest
version.)

To work around this problem, we run
  chmod u+w $HOME/.stm32cubemx/thirdparties/db/updaterThirdParties.xml
when the startup wrapper script recognizes this problem.
